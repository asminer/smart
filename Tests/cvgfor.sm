
/*
   $Id$
*/

/*
 *   Tests a for statement within a converge statement
 */


/*

  x[i] = i

*/

// # Verbose true

# ConvergePrecision 1e-7

print("Linear system using gauss-seidel, by hand\n");
# UseCurrent true

converge {
  print("[");
  for (int i in {1..10}) {
    real x[i] guess i*100;
    real x[i] := cond (i==1, (15-x[10])/5, (6*i-1-x[i-1])/5); 
    print(x[i], cond(i<10, ", ", "")); 
  }
  print("]\n");
}

print("Array x has converged to: [");

for (int i in {1..10}) {
  print(x[i], cond(i<10, ", ", "")); 
}
print("]\n");


print("Linear system using Jacobi, by hand\n");
# UseCurrent false

converge {
  print("[");
  for (int i in {1..10}) {
    real y[i] guess i*100;
    real y[i] := cond (i==1, (15-y[10])/5, (6*i-1-y[i-1])/5); 
    print(y[i], cond(i<10, ", ", "")); 
  } 
  print("]\n");
}

print("Array y has converged to: [");

for (int i in {1..10}) {
  print(y[i], cond(i<10, ", ", "")); 
}
print("]\n");

print("Multiple guess error:\n");
converge {
  for (int i in {8..10}) {
    real w[i] guess i*10;
    real w[i] guess i*10+5;
    real w[i] := sqrt(w[i]);
  }
  for (int i in {8..10}) {
    real v[i] := sqrt(w[i]);
    real v[i] guess i*10;
  }
}

/*
   This is a problem because arrays are *very* tightly coupled
   to their surrounding iterators.
   So, although the below makes sense, it is illegal.
*/
print("Bad split #1:\n");
converge {
  for (int i in {8..10}) {
    real u[i] guess i*10;
  }
  for (int i in {8..10}) {
    real u[i] := sqrt(u[i]);
  }
}

print("Bad split #2:\n");
converge {
  for (int i in {8..10}) {
    real t[i] guess i*10;
  }
}
converge {
  for (int i in {8..10}) {
    real t[i] := sqrt(t[i]);
  }
}
