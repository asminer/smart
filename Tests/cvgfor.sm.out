Linear system using gauss-seidel, by hand
[-197, 41.6, -4.92, 5.584, 4.6832, 6.06336, 6.98733, 8.00253, 8.99949, 10.0001]
[0.99998, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Array x has converged to: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Linear system using Jacobi, by hand
[100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]
[-197, -17.8, -36.6, -55.4, -74.2, -93, -111.8, -130.6, -149.4, -168.2]
[36.64, 41.6, 6.96, 11.92, 16.88, 21.84, 26.8, 31.76, 36.72, 41.68]
[-5.336, -5.128, -4.92, 3.208, 3.416, 3.624, 3.832, 4.04, 4.248, 4.456]
[2.1088, 3.2672, 4.4256, 5.584, 5.1584, 6.3168, 7.4752, 8.6336, 9.792, 10.9504]
[0.80992, 1.77824, 2.74656, 3.71488, 4.6832, 5.96832, 6.93664, 7.90496, 8.87328, 9.8416]
[1.03168, 2.03802, 3.04435, 4.05069, 5.05702, 6.06336, 7.00634, 8.01267, 9.01901, 10.0253]
[0.994931, 1.99366, 2.9924, 3.99113, 4.98986, 5.9886, 6.98733, 7.99873, 8.99747, 9.9962]
[1.00076, 2.00101, 3.00127, 4.00152, 5.00177, 6.00203, 7.00228, 8.00253, 9.00025, 10.0005]
[0.999899, 1.99985, 2.9998, 3.99975, 4.9997, 5.99965, 6.99959, 7.99954, 8.99949, 9.99995]
[1.00001, 2.00002, 3.00003, 4.00004, 5.00005, 6.00006, 7.00007, 8.00008, 9.00009, 10.0001]
[0.99998, 2, 3, 3.99999, 4.99999, 5.99999, 6.99999, 7.99999, 8.99998, 9.99998]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[0.999999, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Array y has converged to: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Multiple guess error:
ERROR in file cvgfor.sm near line 66:
	Duplicate guess for identifier w
ERROR in file cvgfor.sm near line 71:
	Array v already defined
Bad split #1:
ERROR in file cvgfor.sm near line 86:
	Array u already defined
Bad split #2:
ERROR in file cvgfor.sm near line 98:
	Array t already defined
